CriticalMass/Feature.swift

```diff
@@ -0,0 +1,23 @@
+
+//
+//  Feature.swift
+//  CriticalMaps
+//
+//  Created by Felizia Bernutz on 08.12.19.
+//  Copyright © 2019 Pokus Labs. All rights reserved.
+//
+
+import Foundation
+
+private var currentState: [Feature: Bool] = [
+    .friends: false
+]
+
+enum Feature {
+    case friends
+
+    var isActive: Bool {
+        set { currentState[self] = newValue }
+        get { currentState[self] ?? false }
```

Comment:
To me this feels overcomplicated. Why not use a simple `Bool`?
Thing that is bugging me is the nil-coalescing operator for asking the dict

Reply:
Sorry, I don't quite understand what you mean. 
You need something to store the current state for the Feature in, this is `currentState` as private dictionary, so you save the state of each feature in it, if it is activated or not. To get the value if a feature is active or not, you can use `Feature.friends.isActive` which returns a `Bool` and it reads out the value from the dictionary. If the dictionary does not contain the Feature the default for the value is `false`. 
Of course you could make `currentState` public and read out and write in there directly from everywhere, but I prefer the API `Feature.isActive` over `currentState[Feature.friends]`. 

I would be interested, which idea you had in mind. Can you share a code snippet?

Reply:
My idea is to go with something simple like
```swift
enum FeatureFlags {
    static let friends = true
}
```

The proposed solution is fine though. Just want to avoid `??` if possible

Reply:
And the API callside is nicer with an isActive property. 

But maybe that could be a computed property which directly maps the enum cases to bools?

Reply:
```swift
var isActice: Bool {
    switch self {
        case .friends: true
    }
}
```

Reply:
That are nice ideas, they are even simpler! 
A disadvantage with these is though, that you can't change the feature toggle specific for tests. I imagine a snapshot test (for example) which should test a snapshot with activated feature toggle, so when removing the feature toggle the test is still successful. 
Therefore you could add this as test preparation and test with the "new" feature: 

```swift
override func setUp() {
    super.setUp()

    var feature = Feature.friends
    feature.isActive = true
}
```

But maybe we never need these kind of features and we should add these things, when we really need them. So I think I'd go with the `static let` version. What would you think?

Reply:
We already have tests that would require a Feature check (e.g. AppDataStoreTests). 
The Static API looks super nice but makes testing super and I don't like having a global static state that controls the logic. I would prefer passing the state  around with dependency injection like
```swift
featureState.isEnabled(.friends)
```

For tests we could inject a different featureState with the  feature enabled

Reply:
ok I'll see, will try that

Reply:
But that's just my opinion if you both prefer a nicer API that is static that is also fine with me as long as we can somehow enable or disable features for tests

Reply:
Sorry, I don't have much time lately. 🙈If it's fine for you both to start with my current solution, I would be happy if we merge it like that. Feel free to refactor it later, if you want to.

Reply:
No worries! Let’s do that